### PR TITLE
Provide way to debug delta relations and expose iteration counter

### PR DESCRIPTION
**Pull request by julienhenry**
Introducing new way to debug fix point computation on recursive relations: `.debug_delta`.

Example:
```
.decl R(x:number)
.output R()
.decl D = debug_delta(R)
```

This produces a new output by storing each tuples from the `@delta_R` relation after each main fix point loop iteration, adding the iteration number where this tuple was produced.

To obtain the iteration number, I introduce some new RAM nodes (Assign and Variable) to be able to talk about a variable and modify its value. I used these new nodes to introduce a variable storing the iteration number for each LOOP node.

Now that we have an iteration counter for main iteration loops, we can expose this to the user with the `recursive_iteration_cnt()` intrinsic functor. This has some use cases e.g. when someone wants to control the convergence of the fix point loop :

```
R(x + 1) :-
    R(x),
    recursive_iteration_cnt() < 100.
```

### CODE REVIEW

File: src/parser/scanner.ll [hunk 1]
```diff
@@ -208,6 +209,7 @@ WS [ \t\r\v\f]
 "to_string"                           { return yy::parser::make_TOSTRING(yylloc); }
 "to_unsigned"                         { return yy::parser::make_TOUNSIGNED(yylloc); }
 "choice-domain"                       { return yy::parser::make_CHOICEDOMAIN(yylloc); }
+"__ITERATION__"                       { return yy::parser::make_ITERATION(yylloc); }
```
Comment (quentin) [inline src/parser/scanner.ll:212]: Should the iteration counter be made available as an intrinsic functor form like the `autoinc()` ?

File: src/ram/Assign.h [hunk 2]
```diff
@@ -0,0 +1,90 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2021, The Souffle Developers. All rights reserved
```
Comment (quentin) [inline src/ram/Assign.h:3]: `2023` ?

File: src/ram/Variable.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace souffle::ram {
+
+/**
+ * @class Variable
+ * @brief Represents a variable of the RAM
+ *
+ * For example:
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * unsigned(5)
```
Comment (quentin) [inline src/ram/Variable.h:30]: this example does not seem related to the `Variable` class

File: src/ast/IterationCounter.cpp [hunk 4]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2020, The Souffle Developers. All rights reserved
```
Comment (quentin) [inline src/ast/IterationCounter.cpp:3]: `2023` ?

File: src/ast/IterationCounter.cpp [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ * - <souffle root>/licenses/SOUFFLE-UPL.txt
+ */
+
+#include "ast/IterationCounter.h"
+#include <ostream>
+
+namespace souffle::ast {
+
+void IterationCounter::print(std::ostream& os) const {
+    os << "@iteration";
```
Comment (quentin) [inline src/ast/IterationCounter.cpp:15]: Should the printer be as close as possible to the original Souffle input source code?  Maybe `__ITERATION__`.

File: src/ast/IterationCounter.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2021, The Souffle Developers. All rights reserved
```
Comment (quentin) [inline src/ast/IterationCounter.h:3]: `2023`

File: src/ast/analysis/typesystem/TypeConstrainsAnalysis.cpp [hunk 7]
```diff
@@ -174,6 +174,10 @@ void TypeConstraintsAnalysis::visit_(type_identity<Counter>, const Counter& coun
     addConstraint(isSubtypeOf(getVar(counter), typeEnv.getConstantType(TypeAttribute::Signed)));
 }
 
+void TypeConstraintsAnalysis::visit_(type_identity<IterationCounter>, const IterationCounter& counter) {
+    addConstraint(isSubtypeOf(getVar(counter), typeEnv.getConstantType(TypeAttribute::Unsigned)));
```
Comment (quentin) [inline src/ast/analysis/typesystem/TypeConstrainsAnalysis.cpp:178]: I am wondering if a `Signed` would be more appropriate, my guess is that `number`s are more frequent than `unsigned` so it would save some `as` conversions. What do you think?

File: src/parser/scanner.ll (hunk 1 above)
Comment (julienhenry) [reply src/parser/scanner.ll:212]: Yes, I wasn't sure how to name it. Something like "main_loop_iteration()" ?

File: src/ast/analysis/typesystem/TypeConstrainsAnalysis.cpp (hunk 7 above)
Comment (julienhenry) [reply src/ast/analysis/typesystem/TypeConstrainsAnalysis.cpp:178]: Personally I'd prefer keeping the unsigned because it makes more sense for a loop counter.
The way I envision people will use it is by passing it directly to a user defined functor `@my_functor(recursive_loop_cnt(), x, y)` so the functor behaves differently dependending on the loop iteration.
But I don't expect people to use it much in operations involving other numbers.

File: src/parser/scanner.ll (hunk 1 above)
Comment (quentin) [reply src/parser/scanner.ll]: I'm fine with that, the identifier looks long enough to not clash with existing code.

**Comment by quentin** [comment]
As discussed offline, an alternative syntax would be: `.decl IDENT = debug_delta(RELATION)`.

```souffle
.decl R(x:number)
R(x) :- ... x ... R(y).

// declare a relation as being the "debug_delta" of another.
// the new relation has the signature of the original one 
// with one additional tail element that is the iteration counter. 
.decl R_debug = debug_delta(R)
.output R_debug()

.decl use_R_debug(x:number, min_iter:unsigned)
use_R_debug(x,min_iter) :-
  R_debug(x, _),
  min_iter = min iter : R_debug(x, iter).
```

That would be equivalent to:
```
.decl R_debug(x:number, iter:unsigned)
R_debug(x, relation_iteration_cnt()), R(x) :- ... x ... R(y).
```

For a non-recursive rule `R`, the `debug_delta(R)` would contains all facts of `R` and the iteration counter element value `0`?

**Review by quentin — APPROVED** [review]
Interesting debug feature.
Shall we document the new syntax on Souffle's web page?

**Review by quentin — APPROVED** [review]
